p2p/switch.go

```diff
@@ -458,72 +450,70 @@ func (sw *Switch) addInboundPeerWithConfig(conn net.Conn, config *PeerConfig) er
 
 // dial the peer; make secret connection; authenticate against the dialed ID;
 // add the peer.
-func (sw *Switch) addOutboundPeerWithConfig(addr *NetAddress, config *PeerConfig, persistent bool) (Peer, error) {
+func (sw *Switch) addOutboundPeerWithConfig(addr *NetAddress, config *PeerConfig, persistent bool) error {
 	sw.Logger.Info("Dialing peer", "address", addr)
-	peer, err := newOutboundPeer(addr, sw.reactorsByCh, sw.chDescs, sw.StopPeerForError, sw.nodeKey.PrivKey, config, persistent)
+	peerConn, err := newOutboundPeerConn(addr, config, persistent, sw.nodeKey.PrivKey)
 	if err != nil {
 		sw.Logger.Error("Failed to dial peer", "address", addr, "err", err)
-		return nil, err
+		return err
 	}
-	peer.SetLogger(sw.Logger.With("peer", addr))
 
-	err = sw.addPeer(peer)
-	if err != nil {
+	if err := sw.addPeer(peerConn); err != nil {
 		sw.Logger.Error("Failed to add peer", "address", addr, "err", err)
-		peer.CloseConn()
-		return nil, err
+		peerConn.CloseConn()
+		return err
 	}
-	sw.Logger.Info("Dialed and added peer", "address", addr, "peer", peer)
-	return peer, nil
+	return nil
 }
 
 // addPeer performs the Tendermint P2P handshake with a peer
 // that already has a SecretConnection. If all goes well,
 // it starts the peer and adds it to the switch.
 // NOTE: This performs a blocking handshake before the peer is added.
 // NOTE: If error is returned, caller is responsible for calling peer.CloseConn()
-func (sw *Switch) addPeer(peer *peer) error {
-	// Filter peer against white list
-	if err := sw.FilterConnByAddr(peer.Addr()); err != nil {
-		return err
-	}
-
-	// Exchange NodeInfo with the peer
-	if err := peer.HandshakeTimeout(sw.nodeInfo, time.Duration(sw.peerConfig.HandshakeTimeout*time.Second)); err != nil {
-		return err
-	}
+func (sw *Switch) addPeer(pc peerConn) error {
+	addr := pc.conn.RemoteAddr()
+	id := pc.id
 
 	// Avoid self
-	if sw.nodeKey.ID() == peer.ID() {
+	if sw.nodeKey.ID() == id {
 		return ErrSwitchConnectToSelf
 	}
 
 	// Avoid duplicate
-	if sw.peers.Has(peer.ID()) {
+	if sw.peers.Has(id) {
 		return ErrSwitchDuplicatePeer
 	}
 
 	// Filter peer against white list
-	if err := sw.FilterConnByPubKey(peer.PubKey()); err != nil {
+	if err := sw.FilterConnByAddr(addr); err != nil {
+		return err
+	}
+	if err := sw.FilterConnByID(id); err != nil {
```

Comment:
Have you tried running this with `[p2p] auth_enc=false`? If it is false, then id will be `nil`, right? So effectively, filter will not work

Reply:
Hmm crap. I think we need some extra work to make this right for AuthEnc=false. We prob dont want the id to be empty, even if it can't be verified, because we use it eg. as a key for the AddrBook. So if we want AddrBook to work with AuthEnc, we need to do something smarter.

OK thanks